pkgs/development/python-modules/torch/default.nix

```diff
@@ -7,17 +7,9 @@
   config,
   cudaSupport ? config.cudaSupport,
   cudaPackages,
+  cudnn ? null,
   autoAddDriverRunpath,
-  effectiveMagma ?
-    if cudaSupport then
-      magma-cuda-static
-    else if rocmSupport then
-      magma-hip
-    else
-      magma,
```

Comment:
I wanted to limit the override surface of the package and make it easier to override:

- by having a single argument for a package, it's unambiguous which will be used; the alternative is to know to override either `effectiveMagma` or whichever argument `effectiveMagma` selects
- moving the selection logic to the caller gives us more room to express it (since we don't need to put it all inside an expression to the right of a `?`)
- `callPackage` pulls packages from the scope it was constructed with in a strict fashion, so missing packages cause evaluation errors (even if they're unused)
- overrides of `torch` (e.g., `torchWithCuda` or `torchWithoutRocm`) are in `python-packages.nix`, so it made sense to me that the selection logic should belong in the same file (since we use package selection to ensure packages are built with proper versions of their dependencies)

I had thought about trying to do the same in the function argument attribute set in the expression itself, but even if `pkgs` was in scope for me to do the selection argument, I'd still need to use `effective*` values in the attribute set for each of the packages with selection logic. I say that because I'm not aware of a way to tell callPackages not to provide a value for some value: within the expression, I'd have no way of knowing whether arguments like `cudaPackages` were supplied by `callPackage` from `pkgs`, or if they're an override, and therefore no way of knowing whether to set `effectiveCudaPackages` to the value of `cudaPackages` or ignore it and use the package selection logic to get what should be the default value.

I'm also hesitant to use versioned package names; `torch` can build with CUDA 11.8 or 12.1, but it feels odd to require both exist by adding them as arguments to the attribute set fed to `callPackage`. I don't like the idea of picking just one of them either, as `torch.override { cudaPackages_11_8 = cudaPackages_12_1; }` feels bad to me.

What are your thoughts?

Reply:
- Eval errors are prevented by supplying default values
- The point of using `effective*` names is that we know for sure these don't exist in the package sets, so they're either explicitly supplied by the user or computed by us in the default value expression. (This is in response e.g. to "callPackages not to provide a value for some value")
- IIRC the call-site override approach can break with overlays (maybe this is what you refer to with the torch's magma selection logic)

But yes, it's annoying that `callPackage` actually serves multiple different purposes here, and also that it doesn't allow distinguishing between parent scopes at different depths... (to concur with "I'm also hesitant to use versioned package names; torch can build with CUDA 11.8 or 12.1")